test/suite/wthooks.py

```diff
@@ -0,0 +1,256 @@
+# OTHER DEALINGS IN THE SOFTWARE.
+#
+# [TEST_TAGS]
+# ignored_file
+# [END_TAGS]
+#
+# WiredTigerHookManager
+#   Manage running of hooks
+#
+from __future__ import print_function
+
+import abc, os, sys
+from importlib import import_module
+import wiredtiger
+
+# Three kinds of hooks available:
+HOOK_REPLACE = 1     # replace the call with the hook function
+HOOK_NOTIFY = 2      # call the hook function after the function
+HOOK_ARGS = 3        # transform the arg list before the call
+
+# Print to /dev/tty for debugging, since anything extraneous to stdout/stderr will
+# cause a test error.
+def tty(message):
+    from wttest import WiredTigerTestCase
+    WiredTigerTestCase.tty(message)
+
+################
+# Hooks Overview
+#
+# Here are some useful terms to know, with some commentary for each.
+#
+#   API functions
+#      potentially any WiredTiger API functions that a hook creator wishes to modify (like
+#      Session.rename).  In Python most everything is an object.  Of course an instance of
+#      "Session" is an object, but also the "Session" class itself is an object.  The Session.rename
+#      function is also an object (of a certain form that can be called).  Also in Python,
+#      attributes on an object don't have to be "pre-declared", they can be created at any time.
+#      So it's easy to imagine assigning Session._rename_orig to be (the original value of)
+#      Session.rename, and then assigning Session.rename to be some other function object, that
+#      knows how to do something and then perhaps calls Session._rename_orig .  This is the
+#      essence of the hook concept.
+#
+#  Hook Creator:
+#      A way to attach a set of "behavior modifications" to various API functions.  More precisely,
+#      a hook creator derives from WiredTigerHookCreator and sets up a number of "hook functions",
+#      that are actions that are done either just before, after, or instead of, an API function.
+#      A XxxxHookCreator lives in a hook_xxxx.py file.  When a HookCreator is loaded, it may be
+#      given an optional argument.  This argument comes from the original python command line.
+#      For example, "python run.py --hook abc" loads hook_abc.py (where it expects to find a hook).
+#      "python run.py --hook abc=123" loads hook_abc.py with an argument "123".
+#
+#  Hook Function:
+#      One function that will be called before, after or instead of, an API function.  A hook
+#      function will be bound to an API function.  It is the job of the HookCreator to set up that
+#      binding.  It is possible to have multiple hook functions bound to the same API function.
+#      A hook function that replaces an API function will have the same args as the function
+#      it replaces (but there is a trick to give it additional context if needed -
+#      see session_create_replace in hook_demo.py).
+
+# hooked_function -
+# A helper function for the hook manager.
+def hooked_function(self, orig_func, hooks_name, *args):
```

Comment:
By passing in the function, rather than the string, we avoid doing a `hasattr()` or `getattr()` on every call to the function.  Some of the functions may be very high usage (like hooks on any of the `Cursor` methods, especially).

Reply:
oh yeah, that makes alot of sense!